compiler/testData/psi/CallsInWhen_ERR.kt

```diff
@@ -0,0 +1,7 @@
+fun foo() {
+  when (a) {
+    a.foo<T> -> a
```

Comment:
`when` expression allow for writing code like the following:
```kotlin
when {
    a < b, b > c -> TODO()
}
```
Those aren't even unlikely cases. Making a solid suggestion would require a lot more fine grained introspection of all the different cases. IntelliJ autocompletes function calls with `()` at the end, which in turn means that the compiler will consider it as a type argument list. This should strike somewhat of a balance between typing function calls and comparisons.

Reply:
The problem that I see is, that any incomplete expression would never be successfully parsed. But, what we could do the following:

Currently we do this:
```java
if (kind == KotlinParsing.TypeArgumentListKind.FAULTY_TYPE_ARGUMENT_LIST) {
    if (!isAtStrongTypeArgumentListPostfixIndicator()) {
        typeArgumentList.rollbackTo();

        mark().done(TYPE_ARGUMENT_LIST_LIKE_EXPRESSION);
        return false;
    }

    typeArgumentList.rollbackTo();

    if (isAtConditionalExpression()) {
        mark().done(TYPE_ARGUMENT_LIST_LIKE_EXPRESSION);
        return false;
    }

    typeArgumentList = mark();

    myKotlinParsing.tryParseTypeArgumentList();
}
```
We could change it to something like this:
```java
if (kind == KotlinParsing.TypeArgumentListKind.FAULTY_TYPE_ARGUMENT_LIST) {
    int commaCount = getCommaCount(typeArgumentList);

    typeArgumentList.rollbackTo();

    if (isAtConditionalExpression() || !isAtStrongTypeArgumentListPostfixIndicator() && commaCount == 1) {
        mark().done(TYPE_ARGUMENT_LIST_LIKE_EXPRESSION);
        return false;
    }

    typeArgumentList = mark();

    myKotlinParsing.tryParseTypeArgumentList();
}
```
where `getCommaCount(typeArgumentList)` counts the number of commas in the type argument list.

This would change nothing about the correctness, but would lean a lot more towards call suffixes, except when the type argument list contains exactly contains one comma. It wouldn't handle writing expressions like `a < b > false` well, but those are kinda nonsense anyways. And due to line break being a strong indicator, cases like top level `mutableMapOf<String, String>` would still be considered a call suffix despite containing exactly one comma.

Reply:
To be honest, while mostly that trick with `getCommaCount` looks reasonable, I'd prefer to avoid another level of complexity here.
So, I'd suggest to add `->` to the list of weak call postfix comma, so we'd have a conservative solution that preserves the behavior for `a<b> ->` case.
Next time, once we get some complaints about that behavior or would be more decisive to add another layer of recovery logic, we may introduce it.

Reply:
I guess that would make for good compromise. It would hurt the autocompletion of something like the following, but correctness would be still be given. And as soon as there is an error in the type argument list, it would fall back to a comparison again, so it should be ok.
```kotlin
when {
    a < b,
    a > ->TODO()
}
```

Reply:
Sorry, I didn't understand, but may you elaborate what you mean by "that" here?

Reply:
My bad, I meant your idea to add the `->` to the weak indicator set. I initially only considered the strong indicator set, when I proposed the `getCommaCount()` approach, but that would make it lean too much towards call suffixes. The weak indicator set, however, should make for a good compromise. I added the `ARROW` to the weak indicator set and adjusted the tests.